### ISSUE TITLE
Custom DispatchClass

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#11309]
It will be good if we can define custom DispatchClass on top of the existing ones.

This will allow us to have more control on the block space usages. e.g. we can define certain transactions can only use up to x% of the block spaces.

Use case: I am designing a pallet to replace the pallet-scheduler that uses offchain worker + unsigned tx to trigger scheduled call instead of `on_initlaize` hook, to eliminate potential overweight / panic issue. In order to limit of amount block spaces that a scheduled task can use, I would like to define a custom DispatchClass. This will allow me to define limits for custom DispatchClass with `BlockWeights` to ensure scheduled tasks can never use more than x% of block spaces.

**Comment by bkchr** [comment]
> I am designing a pallet to replace the pallet-scheduler that uses offchain worker + unsigned tx to trigger scheduled call instead of `on_initlaize` hook, to eliminate potential overweight / panic issue.

That sounds like an overkill :D Wouldn't having a hook that runs after inherents are applied be enough?

**Comment by xlc** [comment]
That could still overweight / panic and brick the chain? The benefit of unsigned tx is that it is optional to include them. Hooks are mandatory unless something like https://github.com/paritytech/substrate/issues/8740 is implemented.

**Comment by bkchr** [comment]
But in this hook you can check the current weight consumption. `on_initialize` would already have been run and if that would have already used too much weight, you skip running the scheduled call for the current block and try again in the next block. I think overweight is solved with this. Panicking can not be solved by this, but you should not panic anyway :P

**Comment by xlc** [comment]
The main concern is about under estimated weights, which similar to panic, is not something suppose to happen, but still could happen.

Also not just about weights, as well as PoV size, which we don't yet have the ability to benchmark for.

**Comment by bkchr** [comment]
But even for PoV size your offchain solution will not really help. It can still run above the allowed PoV size.

**Comment by xlc** [comment]
It is easier ask collators to ban unsigned tx & disable offchain worker than recover a bad block with overflow PoV size due to a hook.

**Comment by bkchr** [comment]
Okay, I mean I will not prevent you from writing your solution :) 

Regarding the custom dispatch class, do you think that having a variant that holds some u8 or something would be enough?

**Comment by xlc** [comment]
Yes another u8 should be enough. I don’t think anyone will want more than 256 additional dispatch class

**Comment by xlc** [comment]
I feel this is a bit over engineered so closing this.
